# BFS

- Modeling
- Find Neighbors
- Time complexity of Dijktra: O($E*logV$)

## 993. Cousins in Binary Tree
[993. Cousins in Binary Tree](https://leetcode.com/problems/cousins-in-binary-tree/)

In [47]:
from collections import deque
q = deque()
q.append(5)
q.append(6)
print(q)

# doubly linked list: check the existence need O(n)
if 88 in q:
    print("yes")

deque([5, 6])


In [48]:
def isCousins(self, root, x: int, y: int) -> bool:
    if not root:
        return False
    node = (root, float('inf'))
    q = deque([node])
    while q:
        size = len(q)
        par = []
        # add all siblings and cousions in q
        for _ in range(size):
            cur, parent = q.popleft()
            if cur.val == x or cur.val == y:
                par.append(parent)
            if cur.left:
                q.append((cur.left, cur.val))
            if cur.right:
                q.append((cur.right, cur.val))
        # check the parent list
        if len(par) == 2:
             if par[0] != par[1]:
                return True
             return False

## 785. Is Graph Bipartite?
[785. Is Graph Bipartite?](https://leetcode.com/problems/is-graph-bipartite/)

In [4]:
res = [[0 for _ in range(3)] for _ in range(4)]
res

[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]]

In [3]:
def isBipartite(self, graph) -> bool:
    """
    BFS: BFS for all nodes (not connected case)
    visited = [0 for _ in range(n)]
    """
    if not graph:
        return True
    n = len(graph)
    visited = [0 for _ in range(n)]
    for i in range(n):
        if len(graph[i]) == 0 or visited[i] != 0:
            continue
        if not self.bfs(i, graph, visited):
            return False
    return True

def bfs(self, cur, graph, visited):
    q = deque([cur])
    visited[cur] = 1
    while q:
        node = q.popleft()
        # generate neighbors
        for item in graph[node]:
            if visited[item] == 0:
                visited[item] = -1 * visited[node]
                q.append(item)
            if visited[item] != 0 and visited[item] == visited[node]:
                return False
    return True

## 200. Number of Islands
[200. Number of Islands](https://leetcode.com/problems/number-of-islands/)

In [60]:
# BFS
class Solution:
    def numIslands(self, grid) -> int:
        count = 0
        m = len(grid)
        n = len(grid[0])
        directions = [[0, 1],[1, 0],[0,-1],[-1, 0]]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    grid[i][j] = '0'
                    self.bfs(grid, i, j, directions)
                    count += 1
        return count
    
    def bfs(self, grid, i, j, directions):
        q = deque([[i,j]])
        while q:
            row, col = q.popleft()
            for dr, dc in directions:
                r = row + dr
                c = col + dc
                if r >= 0 and r < len(grid) and c >= 0 and c < len(grid[0]) and grid[r][c] == '1':
                    grid[r][c] = '0'
                    q.append([r, c])

## 994. Rotting Oranges
[994. Rotting Oranges](https://leetcode.com/problems/rotting-oranges/)

In [40]:
class Solution:
    def orangesRotting(self, grid) -> int:
        """
        1. Pre-processing: calculate the fresh_num and put rotten num into q
        2. Edge case: no fresh return 0, no rotten return -1
        3. BFS
            generate fresh orange into q
            change grid -> 2: dedup in-place
            step += 1
        4. check fresh_num 
           return step - 1 !!!!
        """
        m = len(grid)
        n = len(grid[0])
        q = deque()
        fresh_num = 0
        # pre-processing: calculate the fresh_num and put rotten num into q
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    q.append([i, j])
                elif grid[i][j] == 1:
                    fresh_num += 1
        
        #edge case
        # no fresh orange
        if fresh_num == 0:
            return 0
        
        # no rotten orange
        if len(q) == 0:
            return -1
        
        # bfs
        step = 0
        directions = [[0, 1], [0, -1], [1, 0],[-1, 0]]
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                row = cur[0]
                col = cur[1]
                
                # generate neighbors
                neighbors = self.findNeighbors(grid, m, n, row, col, directions)
                fresh_num -= len(neighbors)
                for r, c in neighbors:
                    q.append([r, c])
            step += 1
        
        if fresh_num > 0:
            return -1
        
        return step - 1
    
    def findNeighbors(self, grid, m, n, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if r >= 0 and r < m and c >= 0 and c < n and grid[r][c] == 1:
                grid[r][c] = 2 # do not need deduplication (visited)
                neighbors.append([r, c])
        return neighbors

## 286. Walls and Gates
[286. Walls and Gates](https://leetcode.com/problems/walls-and-gates/)

In [52]:
q = deque()
q.append([1,2,3])
q.append([4,5,6])
a, b, c = q.popleft()
print(a, b, c)

1 2 3


In [56]:
class Solution:
    def wallsAndGates(self, rooms) -> None:
        """
        Do not return anything, modify rooms in-place instead.
        
        For each gate call bfs and matain the minimum val to the empty room
        """
        if not rooms:
            return
        m = len(rooms)
        n = len(rooms[0])
        directions = [[1, 0],[-1, 0],[0, 1],[0,-1]]
        q = deque()
        for i in range (m):
            for j in range (n):
                if rooms[i][j] == 0:
                    q.append([i, j])
        visited = [[False for _ in range(n)] for _ in range(m)]
        self.bfs(rooms, visited, q, directions)
    
    def bfs(self, rooms, visited, q, directions):
        while q:
            size = len(q)
            for i in range(size):
                row, col = q.popleft()
                for d in directions:
                    r = row + d[0]
                    c = col + d[1]
                    if r >= 0 and r < len(rooms) and c >= 0 and c < len(rooms[0]):
                        if not visited[r][c] and rooms[r][c] > 0:
                            visited[r][c] = True
                            q.append([r, c])
                            rooms[r][c] = rooms[row][col] + 1

## 490. The Maze
[490. The Maze](https://leetcode.com/problems/the-maze/)

In [25]:
a = [[False for _ in range(3)] for _ in range(3)]
b = [[1,2],[3,4]]
for row, col in b:
    print(row, col)

1 2
3 4


In [31]:
class Solution:
    def hasPath(self, maze, start, destination) -> bool:
        """
        Build graph: 
            Node(State): the position that the ball stop (intermidiate position won't be the node)
            Edge: 4 directions
        
        """
        m = len(maze)
        n = len(maze[0])
        # initial state
        q = deque([(start[0], start[1])])
        visited =  [[False for _ in range(n)] for _ in range(m)]
        visited[start[0]][start[1]] = True
        directions = [[0, 1],[0, -1],[-1, 0],[1, 0]]
        while q:
            cur = q.popleft()
            row = cur[0]
            col = cur[1]
            
            # check destination
            if row == destination[0] and col == destination[1]:
                return True
            
            # get neighbors
            neighbors = self.getNeighbors(maze, row, col, directions)
            for r, c in neighbors:
                if not visited[r][c]:
                    visited[r][c] = True
                    q.append([r, c])
        return False
        
    def getNeighbors(self, maze, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row
            c = col
            # didn't go to this direction before
            # go striaight 
            while r + dr >= 0 and r + dr < len(maze) and c + dc >= 0 and c + dc < len(maze[0]) and maze[r + dr][c + dc] == 0:
                # not visited check here!!!!
                r += dr
                c += dc
            neighbors.append([r, c])
        return neighbors

## 1197. Minimum Knight Moves
[1197. Minimum Knight Moves](https://leetcode.com/problems/minimum-knight-moves/)

In [26]:
# A*
class Solution:
    def minKnightMoves(self, x: int, y: int) -> int:
        from collections import deque, defaultdict 
        from heapq import heapify, heappop, heappush 
        directions = [
            (2, -1), (2, 1), (-2, -1), (-2, 1), (1, 2), (1, -2), (-1, 2), (-1, -2)
        ] 
        
        step = -1 
        visited = set([(0, 0)])  
        # q = deque()  
        # q.append((0, 0))  
        q = [] 
        heappush(q, (self.heuristic(0, 0, x, y), 0, 0, 0))  # protity, cur_x, cur_y, distance traveled 
        
        while q: 
            # cur_size = len(q) 
            # step += 1 
            # for _ in range(cur_size): 
            estimation, cur_x, cur_y, distance = heappop(q) 
            if cur_x == x and cur_y == y: 
                return distance 

            # neighbors  
            for nx, ny in self.successor(cur_x, cur_y, directions): 
                if (nx, ny) not in visited: 
                    visited.add((nx, ny)) 
                    # q.append((nx, ny))  
                    heappush(q, (estimation + self.heuristic(nx, ny, x, y), nx, ny, distance + 1))
                    # if not visited, add, append 
        

    def heuristic(self, x, y, tx, ty): 
        straight = (x - tx)**2 + (y - ty)**2 
        manhaton = abs(x - tx) + abs(y - ty) 
        # return straight 
        return manhaton 

        
    def successor(self, x, y, directions): 
        # return list of tuples [(), () ...] 
        res = [] 
        for dx, dy in directions: 
            nx = x + dx 
            ny = y + dy 
            res.append((nx, ny)) 
        return res

In [24]:
# Time exceed
class Solution:
    def minKnightMoves(self, x: int, y: int) -> int:
        directions = [[-2,-1],[-2,1],[2,-1],[2, 1], [-1,-2], [1,-2], [-1, 2], [1, 2]]
        visited = set([(0, 0)])
        q = deque([(0, 0)])
        move = 0
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                row = cur[0]
                col = cur[1]
                
                if row == x and col == y:
                    return move
                # find neighbors
                neighbors = self.findNeighbors(row, col, directions)
                for nei in neighbors:
                    if nei not in visited:
                        q.append(nei)
                        visited.add(nei)
            move += 1
        return -1
    
    def findNeighbors(self, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            neighbors.append((r, c))
        return neighbors

In [25]:
# Bi-directional BFS
class Solution:
    def minKnightMoves(self, x: int, y: int) -> int:
        
        if (x,y) == (0,0):
            return 0
        visited = {}
        visited[(0,0)] = 0
        
        visitedr = {}
        visitedr[(x,y)] = 0
        
        def get_nbr(a,b):
            return [
                (a+2,b+1),
                (a+2,b-1),
                (a+1,b+2),
                (a+1,b-2),
                (a-2,b+1),
                (a-2,b-1),
                (a-1,b+2),
                (a-1,b-2)
            ]

        q = collections.deque()
        q.append((0,0))
        
        qr = collections.deque()
        qr.append((x,y))
        
        while q:
            node = q.popleft()
            
            for nbr in get_nbr(node[0],node[1]):
                if nbr not in visited:
                    visited[nbr] = 1 + visited[node]
                    q.append(nbr)
                    
                    if nbr in visitedr:
                        return visited[nbr] + visitedr[nbr]
                        
            noder = qr.popleft()
            
            for nbr in get_nbr(noder[0],noder[1]):
                if nbr not in visitedr:
                    visitedr[nbr] = 1 + visitedr[noder]
                    qr.append(nbr)
                    
                    if nbr in visited:
                        return visited[nbr] + visitedr[nbr]


# Implicit Graph

## 773. Sliding Puzzle
[773. Sliding Puzzle](https://leetcode.com/problems/sliding-puzzle/)

- Follw up: thousands of queries:
    - Start and End is interchangable (undirected graph)
    - End is fixed. Start = End, run BFS using a hashmap to record all the moves start to all the configurations.

In [5]:
class Solution:
    def slidingPuzzle(self, board) -> int:
        """
        1. Each configuration of the board represents a Node
        2, Edge: unweighted 
        3. Successor function: 2 or 3 directions to move 0
        
        Implementation:
        BFS
        Hashset: visited
        check if we get to the target Node tuple(1,2,3,4,5,0)
        keep track of the minimum moves
        
        # matrix --> list
          row = idx // c 
          col = idx % c
        """
        m = len(board)
        n = len(board[0])
        end = (1,2,3,4,5,0)
        directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        # Pre-processing input: convert matrix to string
        start = tuple(board[0] + board[1]) # !!! tuple is hashable!
        # print(start)
        visited = set([start])
        q = deque([start])
        moves = -1
        # BFS
        # neis = self.findNeighbors(m, n, start, directions)
        # print(neis)
        while q:
            size = len(q)
            moves += 1
            for i in range(size):
                cur = q.popleft()
                if cur == end:
                    return moves
                # find neighbors
                neighbors = self.findNeighbors(m, n, cur, directions)
                for nei in neighbors:
                    if nei not in visited:
                        q.append(nei)
                        visited.add(nei)
        return -1
    
    def findNeighbors(self, m, n, cur, directions):
        neighbors = []
        cur_list = list(cur)
        # print(cur_list)
        index_zero = cur_list.index(0)
        # print(index_zero)
        row = index_zero // n
        col = index_zero % n
        # print(row, col)
 
        # print(cur_list)
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if r >= 0 and r < m and c >= 0 and c < n:
                new_index = n * r + c
                cur_list[new_index], cur_list[index_zero] = cur_list[index_zero], cur_list[new_index]
                neighbors.append(tuple(cur_list))
                cur_list[new_index], cur_list[index_zero] = cur_list[index_zero], cur_list[new_index]
        return neighbors

## 752. Open the Lock
[752. Open the Lock](https://leetcode.com/problems/open-the-lock/)


- Complexity: $ O(N^2 * A^N + D)$



```
N is Number of dials (4 in our case)
A is number of alphabets (10 in our case -> 0 to 9)
D is the size of deadends.

There are 10 x 10 x 10 x 10 possible combinations => 10^4 => A^N
For each combination, we are looping 4 times (which is N) and in each iteration, there are substring operations ( which is O(N) * constant) => O(4N*constant) => O(4N) => O(NN) => O(N^2)
Total complexity => A^N * N^2, plus D to create the hashset => N^2 * A^N + D
```

In [ ]:
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        """
        State: a str represents a state
        Successor function: +1 / -1 for each digit (8)
        Use a hashset to avoid to go to the deadends
        Use the same hashset to deduplication
        each level -> step
        """
        # edge case
        if target == '0000':
            return 0
        # edd deadends to the hashset
        visited = set(deadends)
        # print(visited)
        if '0000' in visited:
            return -1
        # bfs
        q = deque(['0000'])
        visited.add('0000')
        steps = 0
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                if cur == target:
                    return steps
                # add neighbors
                # neighbors = self.getNeighbors(cur)
                neighbors = self.successor(cur)
                for nei in neighbors:
                    if nei not in visited:
                        q.append(nei)
                        visited.add(nei)
            steps += 1
        return -1
    
    # Works
    def getNeighbors(self, cur):
        neis = []
        for i in range(len(cur)):
            chars = list(cur)
            digit = int(chars[i])
            # + 1
            a = -1 if digit == 9 else digit
            chars[i] = str(a + 1)
            adds = "".join(chars)
            neis.append(adds)
            
            # - 1
            b = 10 if digit == 0 else digit
            chars[i] = str(b - 1)
            minus = "".join(chars)
            neis.append(minus)
        return neis
    
    # Works
    def successor(self, state):  
        res = []  
        state_list = list(state) 
        # state_list = [int(x) for x in state_list] 

        for i in range(len(state_list)):  
            increase = chr(ord(state_list[i]) + 1) if state_list[i] != '9' else '0' 
            state_list[i] = increase 
            res.append(''.join(state_list)) 
            state_list[i] = state[i]   
            
            decrease =  chr(ord(state_list[i]) - 1) if state_list[i] != '0' else '9' 
            state_list[i] = decrease 
            res.append(''.join(state_list)) 
            state_list[i] = state[i]   

        return res

In [23]:
# Concatinate str
cur = '123'
# b = int(cur[1])
# adds = cur[:1] + str(b + 1) + cur[1 + 1:]
# minus = cur[:1] + str(b - 1) + cur[1 + 1:]
# print(adds, minus)

chars = list(cur)
print("list", chars)
ch = int(chars[1])
print("ch", ch)
chars[1] = str(ch + 1)
print("new_list", chars)
new_s = "".join(chars)
new_s

list ['1', '2', '3']
ch 2
new_list ['1', '3', '3']


'133'

In [11]:
digit = 8
digit = 10 if digit == 0 else digit
digit

8

# Topological Sort

## 207. Course Schedule
[207. Course Schedule](https://leetcode.com/problems/course-schedule/)

- graph: E queue: V
- TC = $O(V + E)$
- SC = $O(V + E)$ 

In [12]:
# TC = O(V + E)
class Solution:
    def canFinish(self, numCourses: int, prerequisites) -> bool:
        """
        Topological sort:
        1. build indgree[], build graph
        2. when indgree == 0 --> put into a list/queue
        3. pop the element from list/queue and add to res[]
        4. decrease the indegree of course with dependency
        """
        n = numCourses
        indegree = self.buildIndegree(n, prerequisites)
        graph = self.buildGraph(prerequisites)
        
        zero_indegree = deque()
        for i in range(n):
            if indegree[i] == 0:
                zero_indegree.append(i)
        # print(zero_indegree)
        res = []
        while zero_indegree:
            cur = zero_indegree.popleft()
            res.append(cur)
            # find neighbors
            neighbors = graph[cur]
            for nei in neighbors:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    zero_indegree.append(nei)
        
        # check if all the courses could be taken
        return len(res) == n
            
    
    def buildIndegree(self, n, prerequisites):
        indegree = [0 for _ in range(n)]
        for cur, pre in prerequisites:
            indegree[cur] += 1
        return indegree
    
    def buildGraph(self, prerequisites):
        graph = defaultdict(list)
        for cur, pre in prerequisites:
            graph[pre].append(cur)
        return graph

## 210. Course Schedule II
[210. Course Schedule II](https://leetcode.com/problems/course-schedule-ii/)

In [14]:
class Solution:
    def findOrder(self, numCourses: int, prerequisites):
        n = numCourses
        indegree = self.buildIndegree(n, prerequisites)
        # print(indegree)
        graph = self.buildGraph(prerequisites)
        # print(graph)
        zero_indegree = deque()
        for i in range(n):
            if indegree[i] == 0:
                zero_indegree.append(i)
        res = []
        while zero_indegree:
            cur = zero_indegree.popleft()
            res.append(cur)
            # find neighbors
            neighbors = graph[cur]
            for nei in neighbors:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    zero_indegree.append(nei)

        return res if len(res) == n else []
    
    
    def buildIndegree(self, n, prerequisites):
        indegree = [0 for _ in range(n)]
        for cur, pre in prerequisites:
            indegree[cur] += 1
        return indegree
    
    def buildGraph(self, prerequisites):
        graph = defaultdict(list)
        for cur, pre in prerequisites:
            graph[pre].append(cur)
        return graph

## 953. Verifying an Alien Dictionary
[953. Verifying an Alien Dictionary](https://leetcode.com/problems/verifying-an-alien-dictionary/)

In [18]:
# enumerate
# order = [3,4,'a', 6]
order = "sdahjkfahk"
for idx, ch in enumerate(order):
    print(idx, ch)

0 s
1 d
2 a
3 h
4 j
5 k
6 f
7 a
8 h
9 k


In [15]:
class Solution:
    def isAlienSorted(self, words, order: str) -> bool:
        """
        1. order_dict
        2. compare words[i] and words[i + 1]
           1) letter order
           2) length
        """
        order_dict = {}
        for i in range(len(order)):
            cur = order[i]
            order_dict[cur] = i
        # print(order_dict)
        for i in range(len(words) - 1):
            w1 = words[i]
            w2 = words[i + 1]
            if not self.check(w1, w2, order_dict):
                return False
        return True
    
    def check(self, w1, w2, order_dict):
        j = 0
        while j < min(len(w1), len(w2)):
            c1 = w1[j]
            c2 = w2[j]
            if order_dict[c1] > order_dict[c2]:
                return False
            elif order_dict[c1] < order_dict[c2]:
                return True
            else:
                if j == len(w2) - 1 and len(w1) > len(w2):
                    return False
                j += 1
        return True

## 269. Alien Dictionary
[269. Alien Dictionary](https://leetcode.com/problems/alien-dictionary/)

In [ ]:
# Better
class Solution:
    def alienOrder(self, words) -> str:
        from collections import deque, defaultdict
        """
        1.Linear scan the words list, comapre the words[i] and word[i + 1] to build the graph
        2. Graph: dict (key = letter, val = set()) --> all letters have a node even if some of them don't have neighbors
        3. Run topological sort on graph
        """
        graph = self.buildGraph(words)
        if not graph:
            return ""
        # print("graph", graph)
        return self.topological_sort(graph)
    
    def topological_sort(self, graph):
        indegree = self.buildIndegree(graph)
        # print("indegree", indegree)
        zero_indegree = deque()
        for letter, num in indegree.items():
            if num == 0:
                zero_indegree.append(letter)
        res = []
        while zero_indegree:
            cur = zero_indegree.popleft()
            res.append(cur)
            # find neighbors
            neighbors = graph[cur]
            for nei in neighbors:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    zero_indegree.append(nei)
        return "".join(res) if len(res) == len(graph) else ""
    

    def buildIndegree(self, graph):
        # !! no dependency node --> indegree set to 0
        indegree = {node: 0 for node in graph}
        for node in graph:
            for neighbor in graph[node]:
                indegree[neighbor] += 1
        return indegree
        
    def buildGraph(self, words):
        graph = {}
        # initialize key for all letters
        for word in words:
            for c in word:
                graph[c] = set()
        
        # build graph
        for i in range(len(words) - 1):
            w1 = words[i]
            w2 = words[i + 1]
            j = 0
            while j < len(w1) and j < len(w2):
                c1 = w1[j]
                c2 = w2[j]
                if c1 != c2:
                    graph[c1].add(c2)
                    break
                j += 1
            # invalid case: "abc", "ab"
            if j == len(w2) and len(w1) > len(w2):
                return None
        return graph

In [19]:
class Solution:
    def alienOrder(self, words) -> str:
        from collections import deque, defaultdict
        """
        1.Linear scan the words list, comapre the words[i] and word[i + 1] to build the graph
        2. Run topological sort on graph
        """
        edges = self.getEdges(words)
        # invalid
        if edges == "invalid":
            return ""
        all_letters = self.allLetters(words)
        
        # no dependency
        if not edges:
            return "".join(all_letters)
        
        graph = self.buildGraph(edges)
        indegree = self.buildIndegree(edges)
        # difference of two sets
        dif = all_letters - indegree.keys()
        zero_indegree = deque(dif)
        
        # topological sort
        res = []
        while zero_indegree:
            cur = zero_indegree.popleft()
            res.append(cur)
            # find neighbors
            neighbors = graph[cur]
            for nei in neighbors:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    zero_indegree.append(nei)
        return "".join(res) if len(res) == len(all_letters) else ""
    
    def buildGraph(self, edges):
        graph = defaultdict(list)
        for pre, cur in edges:
            graph[pre].append(cur)
        return graph
    
    def buildIndegree(self, edges):
        indegree = defaultdict(int)
        for pre, cur in edges:
            indegree[cur] += 1
        return indegree
        
    def getEdges(self, words):
        edges = set()
        for i in range(len(words) - 1):
            w1 = words[i]
            w2 = words[i + 1]
            j = 0
            while j < len(w1) and j < len(w2):
                c1 = w1[j]
                c2 = w2[j]
                if c1 != c2:
                    edges.add((c1, c2))
                    break
                j += 1
            # invalid case: "abc", "ab"
            if j == len(w2) and len(w1) > len(w2):
                return "invalid"
        return edges
    
    def allLetters(self, words):
        all_letters = set()
        for word in words:
            for c in word:
                all_letters.add(c)
        return all_letters

# Dijkstra

## 378. Kth Smallest Element in a Sorted Matrix
[378. Kth Smallest Element in a Sorted Matrix](https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/)

In [39]:
class Solution:
    def kthSmallest(self, matrix, k: int) -> int:
        """
        BFS: generate right and buttom elements
        when expanding a node, k -= 1
        when k == 0
        it's the result
        """
        m = len(matrix)
        n = len(matrix[0])
        q = [[matrix[0][0], 0, 0]]
        heapq.heapify(q)
        visited = [[False for _ in range(n)] for _ in range(m)]
        visited[0][0] = True
        directions = [[0, 1], [1, 0]]
        while q:
            cur = heapq.heappop(q)
            val = cur[0]
            r = cur[1]
            c = cur[2]
            k -= 1
            # check if we find the result
            if k == 0:
                return val
            # find neighbors
            neighbors = self.findNeighbors(matrix, m, n, r, c, directions)
            for row, col in neighbors:
                if not visited[row][col]:
                    heapq.heappush(q, [matrix[row][col], row, col])
                    visited[row][col] = True
        return -1
    
    def findNeighbors(self, matrix, m, n, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if r >= 0 and r < m and c >= 0 and c < n:
                neighbors.append([r, c])
        return neighbors

## 505. The Maze II
[505. The Maze II](https://leetcode.com/problems/the-maze-ii/)

In [ ]:
# TC = O(log(n) * m * n)
# SC = O(m*n)
class Solution:
    def shortestDistance(self, maze, start, destination) -> int:
        """
        1. Graph:
            Nodes: the positions that the ball stops 
                   Use a tuple to maintain the cordinate and current distance(distance, row, col)
            Edge: 4 directions
            Weight: the weight of the node is the distance from the neighbor node
        
        2. Uniform Cost search from start to destination (Dijkstra)
        3. The same node might be generated many times if we find a shorter distance
        4. Deduplication: 
            visited[r][c] represent the current shortest distance
            if the distance doesn't change, we won't visited the node anymore
        """
        m = len(maze)
        n = len(maze[0])
        visited = [[float('inf') for _ in range(n)] for _ in range(m)]
        q = [(0, start[0], start[1])]
        visited[start[0]][start[1]] = 0
        heapq.heapify(q)
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        while q:
            cur = heapq.heappop(q)
            dis = cur[0]
            row = cur[1]
            col = cur[2]
            
            # check current node
            if row == destination[0] and col == destination[1]:
                return dis
            
            # generate neighbors
            neighbors = self.findNeighbors(maze, m, n, row, col, dis, directions)
            for r, c, dis in neighbors:
                if visited[r][c] > dis:
                    heapq.heappush(q, (dis, r, c))
                    visited[r][c] = dis
        return -1
        
    def findNeighbors(self, maze, m , n, row, col, dis, directions):
        neighbors = []
        for dr, dc in directions:
            r = row
            c = col
            distance = dis
            while (r + dr) >= 0 and (r + dr) < m and (c + dc) >= 0 and (c + dc) < n and maze[r + dr][c + dc] == 0:
                r += dr
                c += dc
                distance += 1
            neighbors.append((r, c, distance))
        return neighbors

## 499. The Maze III
[499. The Maze III](https://leetcode.com/problems/the-maze-iii/)

In [36]:
visited = [[[float('inf'),""] for _ in range(3)] for _ in range(3)]
visited

print("lu" < "lul")

True


In [37]:
class Solution:
    def findShortestWay(self, maze, ball, hole) -> str:
        """
        Same as Dijkstra solution in Maze 2
        Custom heapq -> distance > path
        Hole is a stop position
        """
        m = len(maze)
        n = len(maze[0])
        visited = [[[float('inf'),""] for _ in range(n)] for _ in range(m)]
        # initial state
        q = [Node(ball[0], ball[1], 0, "")]
        heapq.heapify(q)
        visited[ball[0]][ball[1]] = [0,""]
        directions = [[0, 1, "r"], [0, -1, "l"], [1, 0, "d"],[-1, 0, "u"]]
        res = []
        while q:
            cur = heapq.heappop(q)
            # print(cur.path, cur.distance)
            # check if get to the hole
            if cur.x == hole[0] and cur.y == hole[1]:
                return cur.path
            
            # generate neighbors
            neighbors = self.findNeighbors(maze, m, n, cur, hole, directions)
            for x, y, dis, path in neighbors:
                if visited[x][y][0] > dis or visited[x][y][0] == dis and visited[x][y][1] > path:
                    heapq.heappush(q, Node(x, y, dis, path))
                    visited[x][y][0] = dis
                    visited[x][y][1] = path
                    
        return "impossible"
    
    def findNeighbors(self, maze, m, n, cur, hole, directions):
        neighbors = []
        for dr, dc, p in directions:
            r = cur.x
            c = cur.y
            dis = cur.distance
            path = cur.path
            while (r + dr) >= 0 and (r + dr) < m and (c + dc) >= 0 and (c + dc) < n and maze[r+dr][c+dc] == 0:
                r += dr
                c += dc
                dis += 1
                # check if the ball passes by the hole
                if r == hole[0] and c == hole[1]:
                    neighbors.append((r, c, dis, path + p))
                    break
            neighbors.append((r, c, dis, path + p))
        return neighbors
    
class Node:
    def __init__(self, x, y, distance, path):
        self.x = x
        self.y = y
        self.distance = distance
        self.path = path
    
    def __lt__(self, node2):
        if self.distance == node2.distance:
            return self.path < node2.path
        return self.distance < node2.distance

In [38]:
class Solution:
    def findShortestWay(self, maze, ball, hole) -> str:
        """
        Same as Dijkstra solution in Maze 2
        Custom heapq -> distance > path
        Hole is a stop position
        """
        m = len(maze)
        n = len(maze[0])
        visited = [[float('inf') for _ in range(n)] for _ in range(m)]
        # initial state
        q = [Node(ball[0], ball[1], 0, "")]
        heapq.heapify(q)
        visited[ball[0]][ball[1]] = 0
        directions = [[0, 1, "r"], [0, -1, "l"], [1, 0, "d"],[-1, 0, "u"]]
        res = []
        while q:
            cur = heapq.heappop(q)
            # print(cur.path, cur.distance)
            # check if get to the hole
            if cur.x == hole[0] and cur.y == hole[1]:
                return cur.path
            
            # generate neighbors
            neighbors = self.findNeighbors(maze, m, n, cur, hole, directions)
            for x, y, dis, path in neighbors:
                if visited[x][y] >= dis:
                    heapq.heappush(q, Node(x, y, dis, path))
                    visited[x][y] = dis
                    
        return "impossible"
    
    def findNeighbors(self, maze, m, n, cur, hole, directions):
        neighbors = []
        for dr, dc, p in directions:
            r = cur.x
            c = cur.y
            dis = cur.distance
            path = cur.path
            while r >= 0 and r < m and c >= 0 and c < n and maze[r][c] == 0:
                r += dr
                c += dc
                dis += 1
                # check if the ball passes by the hole
                if r == hole[0] and c == hole[1]:
                    neighbors.append((r, c, dis, path + p))
                    break
            # check if we don't move (or !!!!!)
            if r - dr != cur.x or c - dc != cur.y:
                neighbors.append((r - dr, c - dc, dis - 1, path + p))
        return neighbors
    
class Node:
    def __init__(self, x, y, distance, path):
        self.x = x
        self.y = y
        self.distance = distance
        self.path = path
    
    def __lt__(self, node2):
        if self.distance == node2.distance:
            return self.path < node2.path
        return self.distance < node2.distance

## 787. Cheapest Flights Within K Stops
[787. Cheapest Flights Within K Stops](https://leetcode.com/problems/cheapest-flights-within-k-stops/)

In [4]:
class Solution:
    def findCheapestPrice(self, n: int, flights, src: int, dst: int, k: int) -> int:
        """
        Node: (dis, num, cur_k)
        Neighbors: (num, weight)
        Visited: dict((num, [cur_dis, cur_k]))
        """
        graph = self.buildGraph(flights)
        # print(graph)
        visited = defaultdict(list)
        visited[src] = [0, k + 1]
        pq = [(0, src, k + 1)]
        heapq.heapify(pq)
        while pq:
            cur = heapq.heappop(pq)
            cur_dis = cur[0]
            cur_src = cur[1]
            cur_k = cur[2]
            
            if dst == cur_src:
                return cur_dis
            
            if cur_k > 0:
                neighbors = graph[cur_src]
                # print("neis", neighbors)
                for num, weight in neighbors:
                    new_node = (cur_dis + weight, num, cur_k - 1)
                    if num not in visited:
                        visited[num] = [cur_dis + weight, cur_k - 1]
                        heapq.heappush(pq, new_node)
                    else:
                        prev_dis = visited[num][0]
                        prev_k = visited[num][1]
                        # if prev_dis > cur_dis + weight or prev_k < cur_k - 1:
                        if prev_dis != cur_dis + weight: #or prev_k < cur_k - 1:
                            visited[num] = [cur_dis + weight, cur_k - 1]
                            heapq.heappush(pq, new_node) 

        return -1

    def buildGraph(self, flights):
        # (num, [(des1, weight1), (des2, weight2)])
        graph = defaultdict(list)
        for cur in flights:
            graph[cur[0]].append((cur[1], cur[2]))
        return graph

In [3]:
# Time exceed
class Solution:
    def findCheapestPrice(self, n: int, flights, src: int, dst: int, k: int) -> int:
        """
        Node: (dis, num, cur_k)
        Neighbors: (num, weight)
        Visited: set((cur_dis, cur_src, cur_k))
        """
        graph = self.buildGraph(flights)
        # print(graph)
        visited = set()
        pq = [(0, src, k + 1)]
        heapq.heapify(pq)
        while pq:
            cur = heapq.heappop(pq)
            print("cur", cur)
            visited.add(cur)
            
            cur_dis = cur[0]
            cur_src = cur[1]
            cur_k = cur[2]
            
            if dst == cur_src:
                return cur_dis
            
            if cur_k == 0:
                continue

            if cur_k > 0:
                neighbors = graph[cur_src]
                # print("neis", neighbors)
                for num, weight in neighbors:
                    # print("nums", num)
                    # print("weight", weight)
                    new_node = (cur_dis + weight, num, cur_k - 1)
                    if new_node not in visited:
                        heapq.heappush(pq, new_node)

        return -1

    def buildGraph(self, flights):
        # (num, [(des1, weight1), (des2, weight2)])
        graph = defaultdict(list)
        for cur in flights:
            graph[cur[0]].append((cur[1], cur[2]))
        return graph

## 743. Network Delay Time
[743. Network Delay Time](https://leetcode.com/problems/network-delay-time/)

In [6]:
class Solution:
    def networkDelayTime(self, times, n: int, k: int) -> int:
        """
        Weighted Directed Graph
        BuildGraph -> adjacency list
        Node: (cur_time, num)
        Dijktra --> until all the nodes are visited
        Visited: dict(key = num, value: cur_time)
        check expand a node, add to a final check set() == n
        """
        
        # build graph
        graph = self.buildGraph(times)
        # print(graph)
        # intial state
        pq = [(0, k)]
        heapq.heapify(pq)
        visited = defaultdict(int)
        visited[k] = 0
        received = set()
        while pq:
            cur = heapq.heappop(pq)
            cur_time = cur[0]
            num = cur[1]
            received.add(num)
            
            #check the termination condition
            if len(received) == n:
                return cur_time
            
            # find neighbors
            neighbors = graph[num]
            for nei, weight in neighbors:
                if nei not in visited:
                    visited[nei] = cur_time + weight
                    heapq.heappush(pq, (cur_time + weight, nei))
                else:
                    if cur_time + weight < visited[nei]:
                        visited[nei] = cur_time + weight
                        heapq.heappush(pq, (cur_time + weight, nei))
        return -1
                
    
    def buildGraph(self, times):
        graph = defaultdict(list)
        for u, v, w in times:
            graph[u].append((v, w))
        return graph

## 1293. Shortest Path in a Grid with Obstacles Elimination
[1293. Shortest Path in a Grid with Obstacles Elimination](https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/)

In [1]:
class Solution:
    def shortestPath(self, grid, k: int) -> int:
        """
        Node (i, j, cur_k)
        Call BFS
        ！！！！ cur_k should be in the state
        we might visit the same cell but if the k is different, the path and distance are also different!
        """
        m = len(grid)
        n = len(grid[0])
        target = (m - 1, n - 1)
        q = deque([(0, 0, k, 0)])
        visited = set()
        visited.add((0,0,k))
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        while q:
            cur = q.popleft()
            row = cur[0]
            col = cur[1]
            cur_k = cur[2]
            dis = cur[3]
            
            if (row, col) == target:
                return dis
            
            neighbors = self.findNeighbors(grid, m, n, row, col, directions)
            for r, c in neighbors:
                k_eli = cur_k - grid[r][c]
                if k_eli >= 0:
                    if (r, c, k_eli) not in visited:
                        q.append((r, c, k_eli, dis + 1))
                        visited.add((r, c, k_eli))
        return -1
    
    def findNeighbors(self, grid, m, n, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if 0 <= r < m and 0 <= c < n:
                neighbors.append((r, c))
        return neighbors

## 317. Shortest Distance from All Buildings
[317. Shortest Distance from All Buildings](https://leetcode.com/problems/shortest-distance-from-all-buildings/)

In [8]:
class Solution:
    def shortestDistance(self, grid) -> int:
        space = 0
        buildings = 0
        buildings_idx = set()
        m = len(grid)
        n = len(grid[0])
        # the number of buildings the current position could reach
        canReach = [[0 for _ in range(n)] for _ in range(m)]
        dist = [[0 for _ in range(n)] for _ in range(m)]
        directions = [[0, 1],[0, -1],[1, 0],[-1, 0]]
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    space += 1
                elif grid[i][j] == 1:
                    buildings += 1
                    buildings_idx.add((i, j))
                    is_connected = self.bfs(grid, m, n, i, j, directions, dist, canReach, buildings_idx)
                    if not is_connected:
                        return -1
        if space == 0:
            return -1
        
        min_dist = float('inf')
        for i in range(m):
            for j in range(n):
                if canReach[i][j] == buildings and min_dist > dist[i][j]:
                    min_dist = dist[i][j]
        return -1 if min_dist == float('inf') else min_dist
    
    def bfs(self, grid, m, n, i, j, directions, dist, canReach, buildings_idx):
        q = deque([[i, j]])
        visited = [[False for _ in range(n)] for _ in range(m)]
        visited[i][j] = True
        dis = 0
        build_set = copy.deepcopy(buildings_idx)
        while q:
            size = len(q)
            dis += 1
            for i in range(size):
                cur = q.popleft()
                row = cur[0]
                col = cur[1]
                # neighbors
                neighbors = self.findNeighbors(grid, m, n, row, col, directions, build_set)
                for r, c in neighbors:
                    if not visited[r][c]:
                        q.append([r, c])
                        visited[r][c] = True
                        dist[r][c] += dis
                        canReach[r][c] += 1
        return len(build_set) == 0
        
    
    def findNeighbors(self, grid, m, n, row, col, directions, build_set):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if r >= 0 and r < m and c >= 0 and c < n:
                if grid[r][c] == 0:
                    neighbors.append([r, c])
                if grid[r][c] == 1:
                    if (r, c) in build_set:
                        build_set.remove((r,c))
        return neighbors

# Paths: BFS + DFS

## 797. All Paths From Source to Target
[797. All Paths From Source to Target](https://leetcode.com/problems/all-paths-from-source-to-target/)

In [14]:
class Solution:
    def allPathsSourceTarget(self, graph):
        """
        Adjacency matrix:
        initial: graph[0]
        DFS
        """
        result = [] 
        if not graph:
            return result 
        
        self.dfs(graph, 0, [], result) 
        return result
    
    def dfs(self, graph, cur_node, curRes, result): 
        # terminate condition 
        if cur_node == len(graph) - 1: 
            curRes.append(cur_node) 
            result.append(copy.deepcopy(curRes)) 
            curRes.pop()
            return 
        
        # find neighbors
        curRes.append(cur_node)
        for nei in graph[cur_node]:
            # print(cur)
            # curRes.append(cur_node)
            self.dfs(graph, nei, curRes, result) 
        curRes.pop()

## 127. Word Ladder
[127. Word Ladder](https://leetcode.com/problems/word-ladder/)

In [13]:
# chr() and ord()
a = ord('a')
print(a)
for i in range(1):
    print(chr(a))

97
a


In [ ]:
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        """
        Node: each word in the wordlist
        Edge: differ by a single word
        
        Build Graph: for each digit, modify to 26 letters(exclude itself and check if it exist in the wordlist
        
        BFS1
        """
        word_dict = set(wordList)
        # edge case
        if endWord not in word_dict:
            return 0
        # initial
        q = deque([beginWord])
        visited = set([beginWord])
        moves = 0
        # neis = self.findNeighbors("cog", word_dict)
        # print(neis)
        while q:
            size = len(q)
            moves += 1
            for _ in range(size):
                cur = q.popleft()
                if cur == endWord:
                    return moves
                neighbors = self.findNeighbors(cur, word_dict)
                for nei in neighbors:
                    if nei not in visited:
                        q.append(nei)
                        visited.add(nei)
        return 0
    
    def findNeighbors(self, word, word_dict):
        neighbors = []
        cur = list(word)
        for i in range(len(cur)):
            original = cur[i]
            # print(original)
            a = ord('a')
            for j in range(26):
                new_letter = chr(j + a)
                # print(new_letter)
                if new_letter != original:
                    cur[i] = new_letter
                    new_word = "".join(cur)
                    # print(new_word)
                    if new_word in word_dict:
                        neighbors.append(new_word)
            cur[i] = original
        return neighbors

## 126. Word Ladder II
[126. Word Ladder II](https://leetcode.com/problems/word-ladder-ii/)

## 79. Word Search
[79. Word Search](https://leetcode.com/problems/word-search/)

## 212. Word Search II
[212. Word Search II](https://leetcode.com/problems/word-ladder-ii/)

# Graph

- Unidirected Graph
- Directed Graph
- Weighted Graph
- Bipartite Graph
- Connected Graph
- DAG: directed acyslic Graph
- Circle and Cycle: circle(shape 圈）and Cycle(order 闭环）


- Adjacency List: BFS/DFS $O(V + E)$
- Adjaceny Matrix: BFS/DFS $O(V^2)$

## 133. Clone Graph
[133. Clone Graph](https://leetcode.com/problems/clone-graph/)

In [ ]:
# DFS
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return node
        map_dict = {}
        return self.dfs(node, map_dict)
    
    def dfs(self, node, map_dict):
        # base case
        if not node:
            return node
        
        if node in map_dict:
            return map_dict[node]
        
        # current level
        # create mapping from original to new
        new_node = Node(node.val, [])
        map_dict[node] = new_node
        
        # recursivly clone the neighbors
        # if node.neighbors:
        #     for nei in node.neighbors:
        #         new_nei = self.dfs(nei, map_dict)
        #         new_node.neighbors.append(new_nei)

        new_node.neighbors = [self.dfs(n, map_dict) for n in node.neighbors]

        return new_node

In [20]:
# BFS
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        """
        1.Call BFS to copy all node
        2. for loop to append all neighbors
        
        """
        if not node:
            return node
        
        # q store the nodes that first meet and copy
        q = deque([node])
        map_dict = {}
        new_node = Node(node.val, [])
        map_dict[node] = new_node
        
        # copy node
        while q:
            cur = q.popleft()
            # copy neighbors
            for nei in cur.neighbors:
                if nei not in map_dict:
                    new_nei = Node(nei.val, [])
                    map_dict[nei] = new_nei
                    q.append(nei)
            map_dict[cur].neighbors.append(map_dict[nei])

#         # copy edges
#         for node in map_dict:
#             map_dict[node].neighbors = [map_dict[n] for n in node.neighbors]

        return new_node

## 323. Number of Connected Components in an Undirected Graph
[323. Number of Connected Components in an Undirected Graph](https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/)

In [22]:
class Solution:
    def countComponents(self, n: int, edges) -> int:
        if n <= 1:
            return 1
        graph = self.buildGraph(edges)
        visited = set()
        count = 0
        for i in range(n):
            if i not in visited:
                count += 1
                self.bfs(i, graph, visited)
#                 self.dfs(i, graph, visited)
        return count
    
    # DFS
    def dfs(self, cur, graph, visited):
        # base case
        if cur in visited:
            return
        # current level
        visited.add(cur)
        for nei in graph[cur]:
            self.dfs(nei, graph, visited)
    
    # BFS
    def bfs(self, node, graph, visited): 
        q = deque() 
        
        q.append(node) 
        while q: 
            cur_node = q.popleft()  
            
            for nei in graph[cur_node]: 
                if nei not in visited: 
                    visited.add(nei) 
                    q.append(nei)
    
    def buildGraph(self, edges):
        graph = defaultdict(list)
        for i, j in edges:
            graph[i].append(j)
            graph[j].append(i)
        return graph

## 261. Graph Valid Tree
[261. Graph Valid Tree](https://leetcode.com/problems/graph-valid-tree/)

- Acyclic
- Connected
- Tree: $O(E = V - 1)$

In [23]:
# TC = O(V + E)
# SC = O(V + E)
class Solution:
    def validTree(self, n: int, edges) -> bool:
        """
        1. Tree: E = V - 1
        2. Connected
        """
        if n - 1 != len(edges):
            return False
        graph = self.buildGraph(edges)
        visited = set()
        self.dfs(0, graph, visited)
        return len(visited) == n
    
    def dfs(self, node, graph, visited):
        # base case
        if node in visited:
            return
        visited.add(node)
        for nei in graph[node]:
            if nei not in visited:
                self.dfs(nei, graph, visited)
    
    def buildGraph(self, edges):
        graph = defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        return graph

# Union Find

## 684. Redundant Connection
[684. Redundant Connection](https://leetcode.com/problems/redundant-connection/)

## 2092. Find All People With Secret
[2092. Find All People With Secret](https://leetcode.com/problems/find-all-people-with-secret/)

In [11]:
class Solution:
    """
    Unweighted, directed graph
    1. Sorted the node by time stamp
    2. BuildGraph
       Node : (person: list of [(person1, time1), (person2, time2))
       known: list[] track the earlist time a node is visited
       pq: (time, person)
       !! Only cur_person's known-time is earlier than the neighbors, we could add neighbors to pq!!
    """
    def findAllPeople(self, n: int, meetings, firstPerson: int):
        # Build graph
        graph = self.buildGraph(n, meetings)
        
        # BFS initial
        # Both 0 and firstPerson will spread the secret
        pq = [[0, 0],[0, firstPerson]]
        heapq.heapify(pq)
        
        # Track the earlist time a node is visited
        known = [float('inf') for _ in range(n)]
        
        # BFS
        while pq:
            cur_time, cur_person = heapq.heappop(pq)
            # if the secret already reached the current person
            if known[cur_person] <= cur_time:
                continue
            else:
                known[cur_person] = cur_time
                # find neighbors
                neighbors = graph[cur_person]
                for p, time in neighbors:
                    # Only the current person known the secret eariler than p, the secret could pass to p
                    if known[cur_person] <= time:
                    # if known[p] > time: (Wrong!!! Maybe known[p] = inf, but the meeting happend very early, cur_person haven't know the secret at that time)
                        pq.append([time, p])
        # add all pereson into result
        res = []
        for i in range(n):
            if known[i] < float('inf'):
                res.append(i)
        return res
    
    def buildGraph(self, n, meetings):
        graph = defaultdict(list)
        for p1, p2, time in meetings:
            graph[p1].append([p2, time])
            graph[p2].append([p1, time])
        return graph

# Others

## 417. Pacific Atlantic Water Flow
[417. Pacific Atlantic Water Flow](https://leetcode.com/problems/pacific-atlantic-water-flow/)

## 543. Diameter of Binary Tree
[543. Diameter of Binary Tree](https://leetcode.com/problems/diameter-of-binary-tree/)

## 1245. Tree Diameter
[1245. Tree Diameter](https://leetcode.com/problems/tree-diameter/)

## 1192. Critical Connections in a Network
[1192. Critical Connections in a Network](https://leetcode.com/problems/critical-connections-in-a-network/)

## 399. Evaluate Division
[399. Evaluate Division](https://leetcode.com/problems/evaluate-division/)

## 329. Longest Increasing Path in a Matrix
[329. Longest Increasing Path in a Matrix](https://leetcode.com/problems/longest-increasing-path-in-a-matrix/)

## 489. Robot Room Cleaner
[489. Robot Room Cleaner](https://leetcode.com/problems/robot-room-cleaner/)

## Robot Path Planning

In [6]:
visited = [[False for _ in range(2)] for _ in range(3)]
dir = [[0, 1],[1, 0], [0, -1], [-1, 0]]
dir

[[0, 1], [1, 0], [0, -1], [-1, 0]]

In [43]:
from collections import deque
def plan(grid):
    # corner case
    if not grid:
        return ""
    # find the start position
    m = len(grid)
    n = len(grid[0])
    q = deque()
    visited = [[False for _ in range(n)] for _ in range(m)]
    direction = [[0, 1],[0, -1],[1, 0],[-1, 0]]
    d_char = ['E','W','S','N']
    for i in range(m):
        for j in range(n):
            if grid[i][j] == 'S':
                q.append((i, j, ""))
                visited[i][j] = True
                res = bfs(grid, q, visited, direction, d_char)
    return res

def bfs(grid, q, visited, direction, d_char):
    while q:
        cur = q.popleft()
        row = cur[0]
        col = cur[1]
        path = cur[2]
        
        # desitination
        if grid[row][col] == 'G':
            return path
        # build path with it's neighbors
        for i in range(len(direction)):
            r = row + direction[i][0]
            c = col + direction[i][1]
            if r >= 0 and r < len(grid) and c >= 0 and c < len(grid[0]):
                if not visited[r][c] and grid[r][c] != 'O':
                    # generated
                    visited[r][c] = True
                    q.append((r, c, path+d_char[i]))
    return ""
                    
    
a = [['O','E','E','O','E','E'],
     ['O','E','O','O','E','O'],
     ['S','E','E','E','E','E'],
     ['O','E','O','O','E','E'],
     ['O','E','E','E','E','O']]
print(plan(a))

ESSEE


## 1263. Minimum Moves to Move a Box to Their Target Location
[1263. Minimum Moves to Move a Box to Their Target Location](https://leetcode.com/problems/minimum-moves-to-move-a-box-to-their-target-location/)

## 332. Reconstruct Itinerary
[332. Reconstruct Itinerary](https://leetcode.com/problems/reconstruct-itinerary/)